### PR TITLE
fix(csv): align header-only analysis across engines

### LINKED ISSUE #558: analyze_csv_file drops declared columns and the quality block for a header-only CSV; every engine reports them
## Summary

`analyze_csv_file` / `analyze_csv_from_reader` drop every declared column for a
header-only CSV, and with them the entire quality block. Every engine path
reports the columns. The drift survives serialization, so a saved report omits
two declared columns and its whole `quality` object depending only on which
entry point produced it.

`analyze_csv_file` is public API, re-exported from the facade at
`crates/dataprof/src/lib.rs:35`.

## Reproduction

`header_only.csv`:

```csv
name,age
```

Measured on `master` (unchanged by #557), raw attributes:

```
analyze_csv_file:  cols=[]                quality=None
Incremental:       cols=["name", "age"]   quality=Some((0.0, "Exact"))
Columnar:          cols=["name", "age"]   quality=Some((0.0, "Exact"))
Auto:              cols=["name", "age"]   quality=Some((0.0, "Exact"))
```

And after `serde_json` round-trip, which is what makes this more than an
attribute-level quibble:

```
analyze_csv_file:  serialized column_profiles=0  quality_present=false
Incremental:       serialized column_profiles=2  quality_present=true   ["name","age"]
```

## Root cause

`analyze_csv_from_reader_with_hints` reads and validates the header, then builds
its `StreamingColumnCollection` without pre-registering those names. Column
state is only ever created by `process_record`, so a file with zero data rows
produces zero columns, and `profiles_from_streaming_with_hints` iterates an
empty collection.

The incremental engine does not have this bug because it already pre-registers
the header immediately after `validate_unique_column_names`
(`crates/dataprof-engines/src/streaming/incremental.rs:168`):

```rust
column_stats.init_columns(&names);
```

## Proposed fix

Mirror the engine in `crates/dataprof-csv/src/lib.rs`:

```rust
    let mut column_stats = StreamingColumnCollection::new().with_semantic_hints(semantic_hints);
    // The header declares the columns; a file with zero data rows still has them.
    column_stats.init_columns(&header_names);
    let mut rows_read = 0;
```

Verified locally: `analyze_csv_file` comes into line with all three engines, and
`cargo test --workspace --all-features` is green except one test that pins the
current behavior and needs its assertion updated:

- `crates/dataprof-csv/src/lib.rs:567` `test_analyze_csv_file_empty_file`
  asserts `report.column_profiles.len() == 0` for input `"name,age
"`.

Two existing special cases become redundant and can be deleted in the same
change:

- `crates/dataprof-partial/src/lib.rs:809` — the
  `if columns.is_empty() && !headers.is_empty()` patch in `analyze_structure_csv`
- `crates/dataprof-partial/src/lib.rs:954` — `empty_structure_columns_from_headers`

That also collapses the "no observations means `String`" rule back to a single
definition (`infer_data_type_streaming`) instead of the three places it is
currently spelled out.

## Decision this fix forces

Once the columns exist, `analyze_csv_file` starts emitting a quality
assessment for a header-only file, and the engines' current value for that case
is **overall score `0.0` with confidence `Exact`**. A `0.0` score reads as "this
data is terrible" when the truth is "there are no rows to assess", which is the
`None`-means-not-analyzed versus empty-means-nothing-found distinction from
AGENTS.md, and squarely 0.11's scoring-honesty theme. Two options:

1. Land the parity fix as-is, inheriting `0.0 / Exact`, and file the zero-row
   score question separately.
2. Decide zero-row quality in this issue — most likely `quality: None` or an
   explicit low-confidence marker — and apply it to the engines at the same
   time, since they have the same behavior today.

Option 1 keeps this a one-behavior-change PR; option 2 avoids shipping a score
we would immediately revisit. Worth a call before someone picks this up.

## Relation to other issues

- Follow-up to #557 (fixes #552). #557 aligns `profile`, `infer_schema`, and
  `analyze_structure` for header-only CSV but leaves the shared CSV parser
  entry point out of alignment, so after it merges the parser path is the lone
  dissenter.
- **Not** blocked on #547. That issue is scoped to drift where raw attributes
  disagree but serialized reports agree; this case breaks the serialized
  surface too, so it is a bug under either resolution of the contract question.

## Milestone

Filed for **0.11.0**, since #557 is milestoned 0.11.0 and therefore #552 ships
in 0.11 regardless of its stale 0.12.0 label. Shipping a release where three
surfaces agree and the public parser entry point does not seems worse than the
one-line cost of finishing it. Move to 0.12.0 if option 2 above is chosen and
the zero-row quality decision needs to sit with #547 and the rest of the
contract work.

## Definition of done

- `analyze_csv_file` and every engine report identical columns for a
  header-only CSV, raw and serialized.
- A test asserts that parity across paths rather than asserting zero columns
  for one of them.
- The duplicated header-fallback helpers in `dataprof-partial` are gone, or the
  issue states why they are still needed.

### PR DESCRIPTION
**Pull request by floze-the-genius**
# Pull Request

## 📝 Summary

Make the CSV profiling paths preserve declared columns for a header-only file, so standard, auto, incremental, columnar, and async-streaming analysis all expose the same schema. This also removes the now-redundant partial-analysis fallbacks added before the parser owned that invariant.

## 🔧 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 🚨 Breaking change (would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] ⚡ Performance improvement
- [ ] 🎨 Code refactoring (no functional changes)

## 📋 Changes Made

- Pre-register CSV header names in `StreamingColumnCollection` before reading records in both file-based and async-streaming paths.
- Split the former “empty file” regression into genuinely empty and header-only cases.
- Remove duplicate header-only schema/structure fallbacks from `dataprof-partial` and pin the resulting structure fields.
- Add a cross-engine and serialization regression for header-only CSV input.

**Related Issue:** Closes #558

## 🧪 Testing

- [x] Tested locally with `cargo test`
- [x] Added new tests for this feature
- [x] All existing tests pass
- [x] Tested with sample data files
- [x] Tested edge cases (empty files, large files, etc.)

### Test Commands Run

```bash
cargo test --workspace --all-features
cargo test -p dataprof-engines --features async-streaming
cargo test --features async-streaming --test async_streaming
cargo clippy -p dataprof-csv -p dataprof-partial -p dataprof --all-targets --all-features -- -D warnings
cargo clippy --features async-streaming --all-targets -- -D warnings
cargo fmt --all -- --check
uv sync
uv run maturin develop
uv run pytest python/tests/test_python_api.py -q  # 323 passed
uv run ruff check .
uv run ty check
git diff --check
```

The repository-wide `uv run ruff format --check .` also reports eight pre-existing documentation/skill files that current Ruff would reformat; this PR changes only Rust files.

## 📖 Documentation

- [ ] Updated README.md if user-facing changes
- [x] Added/updated inline code comments
- [ ] Updated docs/ folder if architectural changes
- [x] No documentation changes needed

## ⚡ Performance Impact

- [x] No performance impact
- [ ] Performance improved
- [ ] Performance may be affected (explain below)

## 🚨 Breaking Changes

- [x] No breaking changes

## 📌 Additional Notes

The separate zero-row quality aggregation decision remains tracked in #571 and is intentionally unchanged here.

## ✅ Checklist

- [x] Code builds without warnings
- [x] No linting errors (`cargo clippy`)
- [x] Code is formatted (`cargo fmt`)
- [x] All tests pass
- [x] Self-reviewed my code
- [x] Comments added for complex logic
- [x] No hardcoded secrets or sensitive data

### CODE REVIEW

File: crates/dataprof-csv/src/lib.rs [hunk 1]
```diff
@@ -281,6 +281,8 @@ pub fn analyze_csv_from_reader_with_hints<R: Read>(
     };
 
     let mut column_stats = StreamingColumnCollection::new().with_semantic_hints(semantic_hints);
+    // A header declares the columns even when no data records follow it.
+    column_stats.init_columns(&header_names);
```
[the remark this answers is not in the document]
Comment (floze-the-genius) [reply crates/dataprof-csv/src/lib.rs:285]: Fixed in 37e7498. The async collection now pre-registers the validated header vector before reading data chunks, so header-only CSV streams preserve ordered String profiles with zero counts. I updated the engine regression and added a public Profiler::profile_stream regression. The full workspace/all-features suite, 39 async engine tests, 14 async facade tests, focused cross-engine serialization regression, fmt, diff check, and async Clippy with warnings denied all pass.

**Comment by AndreaBozzo** [comment]
Reviewed and verified: header-only CSV now agrees across `analyze_csv_file`, all three engines, and the async stream, raw and serialized. I reverted each `init_columns` call separately to confirm both parts of the fix are independently covered by the tests you added. Full workspace suite, clippy, fmt and the Python suite are green here.

I pushed two small commits rather than sending this back for another round, since we want it in:

- `schema_from_streaming_stats` now reads the schema off the profiles the parser already returns, instead of looking each header up and `filter_map`ping away a miss. Same result for both callers, but a declared column can no longer be dropped without a signal, and the "no observations means String" rule stays in one place.
- The empty-file early return reports 0 columns (`header_names.len()` is provably 0 once every header is pre-registered), and the empty-file quality assertion now says it records this path's behaviour rather than a settled contract.

Two follow-ups came out of the review, both out of scope here and neither caused by this PR:

- #573 the empty-file half of the same drift: `analyze_csv_file` returns `quality: None` where every engine returns a block scoring 0.0/Exact.
- #574 the async reader accepts duplicate CSV headers and merges two columns into one profile, where the sync parser and the incremental engine both refuse.

Thanks, this was a clean fix and the cross-engine plus serialization regression is exactly the right shape.

**Review by AndreaBozzo — APPROVED** [review]
